src/compiler/checker.ts

```diff
@@ -21330,11 +21370,20 @@ namespace ts {
             return result;
         }
 
+        function getApparentIntersectionTypes(type: IntersectionType): Type[] {
+            const apparent = getApparentType(type);
+            if (apparent.flags & TypeFlags.Intersection) {
+                return (apparent as IntersectionType).types;
+            }
+            return [apparent];
+        }
+
         // Returns the String, Number, Boolean, StringLiteral, NumberLiteral, BooleanLiteral, Void, Undefined, or Null
         // flags for the string, number, boolean, "", 0, false, void, undefined, or null types respectively. Returns
         // no flags for all other types (including non-falsy literal types).
         function getFalsyFlags(type: Type): TypeFlags {
             return type.flags & TypeFlags.Union ? getFalsyFlagsOfTypes((type as UnionType).types) :
+                type.flags & TypeFlags.Intersection ? reduceLeft(getApparentIntersectionTypes(type as IntersectionType), (memo, type) => memo | getFalsyFlags(type), 0 as TypeFacts) :

```

Comment:
Small change to falsy flag calculation to pick up `T & undefined` as falsy when it doesn't reduce to `never`.